consensus/hotstuff/verification/combined_verifier_v3.go

```diff
@@ -0,0 +1,175 @@
+// TODO: return error only, because when the sig is invalid, the returned bool
+// can't indicate whether it's staking sig was invalid, or beacon sig was invalid.
+func (c *CombinedVerifierV3) VerifyVote(signer *flow.Identity, sigData []byte, block *model.Block) (bool, error) {
+
+	// create the to-be-signed message
+	msg := MakeVoteMessage(block.View, block.BlockID)
+
+	sigType, sig, err := signature.DecodeSingleSig(sigData)
+	if err != nil {
+		return false, fmt.Errorf("could not decode signature: %w", modulesig.ErrInvalidFormat)
+	}
+
+	switch sigType {
+	case hotstuff.SigTypeStaking:
+		// verify each signature against the message
+		stakingValid, err := signer.StakingPubKey.Verify(sig, msg, c.stakingHasher)
+		if err != nil {
+			return false, fmt.Errorf("internal error while verifying staking signature: %w", err)
+		}
+		if !stakingValid {
+			return false, fmt.Errorf("invalid staking sig")
+		}
+	case hotstuff.SigTypeRandomBeacon:
+		dkg, err := c.committee.DKG(block.BlockID)
+		if err != nil {
+			return false, fmt.Errorf("could not get dkg: %w", err)
+		}
+
+		// if there is beacon share, there must be beacon public key
+		beaconPubKey, err := dkg.KeyShare(signer.NodeID)
+		if err != nil {
+			return false, fmt.Errorf("could not get random beacon key share for %x: %w", signer.NodeID, err)
+		}
+
+		beaconValid, err := beaconPubKey.Verify(sig, msg, c.beaconHasher)
+		if err != nil {
+			return false, fmt.Errorf("internal error while verifying beacon signature: %w", err)
+		}
+
+		if !beaconValid {
+			return false, fmt.Errorf("invalid beacon sig")
+		}
+	default:
+		return false, fmt.Errorf("invalid signature type %d: %w", sigType, modulesig.ErrInvalidFormat)
+	}
+
+	return true, nil
+}
+
+// VerifyQC verifies the validity of a combined signature on a quorum certificate.
+func (c *CombinedVerifierV3) VerifyQC(signers flow.IdentityList, sigData []byte, block *model.Block) (bool, error) {
+
+	dkg, err := c.committee.DKG(block.BlockID)
+	if err != nil {
+		return false, fmt.Errorf("could not get dkg data: %w", err)
+	}
+
+	// unpack sig data using packer
+	blockSigData, err := c.packer.Unpack(block.BlockID, signers.NodeIDs(), sigData)
+	if err != nil {
+		return false, fmt.Errorf("could not split signature: %w", modulesig.ErrInvalidFormat)
+	}
+
+	msg := MakeVoteMessage(block.View, block.BlockID)
+
+	// verify the beacon signature first since it is faster to verify (no public key aggregation needed)
+	beaconValid, err := dkg.GroupKey().Verify(blockSigData.ReconstructedRandomBeaconSig, msg, c.beaconHasher)
+	if err != nil {
+		return false, fmt.Errorf("internal error while verifying beacon signature: %w", err)
+	}
+	if !beaconValid {
+		return false, nil
+	}
+
+	// verify the aggregated staking and beacon signatures next (more costly)
+
+	verifyAggregatedSignature := func(pubKeys []crypto.PublicKey, aggregatedSig crypto.Signature, hasher hash.Hasher) (bool, error) {
+		keysAggregator, err := modulesig.NewPublicKeyAggregator(pubKeys)
```

Comment:
What is implemented here works, however it doesn't need the `PublicKeyAggregator` stateful API. What you implemented could be done more easily using [this stateless API](https://github.com/onflow/flow-go/blob/master/crypto/bls_multisig.go#L147) for instance. 

We need the stateful structure `PublicKeyAggregator` in order to optimize the key aggregation: the structure keeps internally the latest aggregated key, and adjusts it based on the delta of the latest signers and new signers and gets the new aggregated key. 
For this to work, we need only a single `PublicKeyAggregator` per epoch per set of keys (so one struct for staking keys and another for beacon keys). 

I thought we can add this in another PR since it requires more refactoring (that's why my PR comment only mentioned adding a TODO comment), but feel free to add it here if you think the changes are manageable.

Reply:
If we're planning to make the changes outside this PR, I suggest to not use `PublicKeyAggregator` and use [the simple aggregation](https://github.com/onflow/flow-go/blob/master/crypto/bls_multisig.go#L147)

Reply:
That is exactly what I was trying to bring on the meeting. Since we don't have a clear strategy how to deal per epoch caching I have just used it this way. I didn't know about stateless API. 

I would say let's go with stateless API and once we have a clear understanding how to handle changing epochs and where to do the caching then do the optimization.

Reply:
That works for me, let's just use the stateless API and leave a TODO comment. 
I'll create a ticket to integrate `PublicKeyAggregator` properly.

Reply:
https://github.com/dapperlabs/flow-go/issues/6016

Reply:
The stateless API is already in place